packages/world/src/World.sol

```diff
@@ -0,0 +1,422 @@
+
+    // StoreCore handles checking for existence
+    StoreCore.registerSchema(tableRouteId, schema);
+  }
+
+  /**
+   * Register a schema for a given table id
+   * This overload exists to conform to the Store interface,
+   * but it requires the caller to register a route using `registerRoute` first
+   */
+  function registerSchema(uint256 tableId, Schema schema) public override {
+    // Require caller to own the given tableId
+    if (OwnerTable.get(bytes32(tableId)) != msg.sender)
+      revert RouteAccessDenied(RouteTable.get(bytes32(tableId)), msg.sender);
+
+    // Register the schema
+    StoreCore.registerSchema(tableId, schema);
+  }
+
+  /**
+   * Register a hook for a given table route
+   */
+  function registerTableHook(string calldata tableRoute, IStoreHook hook) public {
+    registerStoreHook(uint256(keccak256(bytes(tableRoute))), hook);
+  }
+
+  /**
+   * Register a hook for a given table route id
+   * This overload exists to conform with the `IStore` interface.
+   */
+  function registerStoreHook(uint256 tableId, IStoreHook hook) public override {
+    // Require caller to own the given tableId
+    if (OwnerTable.get(bytes32(tableId)) != msg.sender)
+      revert RouteAccessDenied(RouteTable.get(bytes32(tableId)), msg.sender);
+
+    // Register the hook
+    StoreCore.registerStoreHook(tableId, hook);
+  }
+
+  /**
+   * Register a hook for a given system route
+   */
+  function registerSystemHook(string calldata systemRoute, ISystemHook hook) public {
+    // TODO
+  }
+
+  /**
+   * Register a system at the given route
+   */
+  function registerSystem(
+    string calldata baseRoute,
+    string calldata systemRoute,
+    System system,
+    bool publicAccess
+  ) public returns (bytes32 systemRouteId) {
+    // Require the system to not exist yet
+    if (SystemRouteTable.has(address(system))) revert SystemExists(address(system));
+
+    // Require the caller to own the base route
+    bytes32 baseRouteId = keccak256(bytes(baseRoute));
+    if (OwnerTable.get(baseRouteId) != msg.sender) revert RouteAccessDenied(baseRoute, msg.sender);
+
+    // Register system route
+    systemRouteId = registerRoute(baseRoute, systemRoute);
+
+    // Store the system address in the system table
+    SystemTable.set({ routeId: systemRouteId, system: address(system), publicAccess: publicAccess });
+
+    // Store the system's route in the SystemToRoute table
+    SystemRouteTable.set({ system: address(system), routeId: systemRouteId });
+
+    // Give the system access to its base route
+    RouteAccessTable.set({ routeId: baseRouteId, caller: address(system), access: true });
+  }
+
+  /**
+   * Grant access to a given route
+   */
+  function grantAccess(string calldata route, address grantee) public {
+    // Require the caller to own the route
+    bytes32 routeId = keccak256(bytes(route));
+    if (OwnerTable.get(routeId) != msg.sender) revert RouteAccessDenied(route, msg.sender);
```

Comment:
A small issue I have with that is usually we need the `routeId` later in the logic, so either we compute the `routeId` twice (once in the main method, once in the helper), or we pass in the `routeId` to the helper directly, which then requires us to grab the `route` pre-image from the `RouteTable` to be able to emit it as part of the error. (In some cases we only have `routeId`, eg `registerSchema`, so we'd need the helper that works with `routeId` anyways.)

The other thing is if we revert in the helper method, it's slightly less clear from looking at the method what the possible outcomes are (because you have to check the helper method first to see that the method can revert with a certain error). Since the check here is pretty simple and readable, I have a slight preference for throwing the error explicitly in the main function.